src/compiler/utilities.ts

```diff
@@ -1952,10 +2045,31 @@ namespace ts {
         return token >= SyntaxKind.FirstAssignment && token <= SyntaxKind.LastAssignment;
     }
 
-    export function isExpressionWithTypeArgumentsInClassExtendsClause(node: Node): boolean {
-        return node.kind === SyntaxKind.ExpressionWithTypeArguments &&
-            (<HeritageClause>node.parent).token === SyntaxKind.ExtendsKeyword &&
-            isClassLike(node.parent.parent);
+    /**
+      * Tests whether the node is an ExpressionWithTypeArguments node that is part of the `extends` 
+      * clause of a class.
+      * @param node The node to test
+      * @param getAncestorOrSelf A callback used to get the ancestor of the starting node, used 
+      * only when traversing the ancestors of the current node in the emitter.
+      * @param offset The offset in the node stack used to get the ancestor of the current node
+      * from the emitter's node stack.
+      * @remarks
+      * The emitter tracks the parent of the current node as it descends into the source file, 
+      * so that we can properly emit synthesized nodes that may not have parent pointers.
+      * We can call `getAncestorOrSelf` with an offset that specifies how far back in the node's 
+      * ancestry to retrieve a parent, grandparent, etc. An offset of zero (0) refers to the 
+      * current node on the top of the node stack, an offset of one (1) refers to its parent, an 
+      * offset of two (2) refers to the grandparent, and so on.
+      */
+    export function isExpressionWithTypeArgumentsInClassExtendsClause(node: Node, getAncestorOrSelf?: (offset: number) => Node, offset?: number): boolean {
+        if (node.kind === SyntaxKind.ExpressionWithTypeArguments) {
+            let parent = getAncestorOrSelf ? getAncestorOrSelf(offset + 1) : node.parent;

```

Comment:
This line/pattern has been repeated many times. Can we consolidate it in a new helper function, like `getParent(...)`?